### ISSUE TITLE
crash while undoing

### ISSUE THREAD

**Issue by JoeyEdmo** [opentoonz/opentoonz#4940]
### Description

was using control z to undo making a frame. Was pressing it a bunch of times in a row. then it crashed. 

![image](https://github.com/opentoonz/opentoonz/assets/97484115/c1e84d0e-d0d7-43af-9a1d-4652c04c51ab)

repeated it here, I dont know why it happens exactly but it's tied to undo. difficulty replicating.

![image](https://github.com/opentoonz/opentoonz/assets/97484115/ca87f243-9255-43e5-98b4-ea2d84715390)

### Steps to Reproduce

unfortunately have not been able to figure out how to remake it, but both times the undo command was involved. 

### Expected Behavior

my previous action is undone and it doesn't crash. 

### Screenshots, Video & Crash Logs

OpenToonz 1.7.1  (Build Mar 14 2023)

Report Date: 2023-06-08 01:38:39
Crash Reason: EXCEPTION_ACCESS_VIOLATION

Build ABI: x86_64-little_endian-llp64
Operating System: Windows 10 Version 2009
OS Kernel: 10.0.19045
CPU Threads: 12

GPU Vendor: NVIDIA Corporation
GPU Model: NVIDIA GeForce RTX 3070 Ti/PCIe/SSE2
GPU Version: 4.6.0 NVIDIA 516.94

Crash File: C:\Users\joeye\AppData\Local\OpenToonz\OpenToonz\crash\Crash-20230608-013839.log
Mini Dump File: C:\Users\joeye\AppData\Local\OpenToonz\OpenToonz\crash\Crash-20230608-013839.dmp

Application Dir: C:/Program Files/OpenToonz
Stuff Dir: C:\OpenToonz stuff

Project Name: sandbox
Scene Name: garbage
Project Path: C:\OpenToonz stuff\sandbox\sandbox_otprj.xml
Scene Path: C:\OpenToonz stuff\sandbox\scenes\garbage.tnz

==== Modules ====
C:\Program Files\OpenToonz\OpenToonz.exe
C:\Windows\SYSTEM32\ntdll.dll
C:\Windows\System32\KERNEL32.DLL
C:\Windows\System32\KERNELBASE.dll
C:\Windows\SYSTEM32\apphelp.dll
C:\Windows\System32\USER32.dll
C:\Windows\System32\win32u.dll
C:\Windows\System32\GDI32.dll
C:\Windows\SYSTEM32\OPENGL32.dll
C:\Windows\System32\gdi32full.dll
C:\Windows\System32\msvcrt.dll
C:\Windows\System32\msvcp_win.dll
C:\Windows\System32\ADVAPI32.dll
C:\Windows\System32\sechost.dll
C:\Windows\System32\ucrtbase.dll
C:\Windows\System32\RPCRT4.dll
C:\Windows\System32\combase.dll
C:\Windows\System32\SHELL32.dll
C:\Windows\System32\ole32.dll
C:\Windows\System32\OLEAUT32.dll
C:\Program Files\OpenToonz\Qt5SerialPort.dll
C:\Program Files\OpenToonz\Qt5Xml.dll
C:\Program Files\OpenToonz\Qt5PrintSupport.dll
C:\Windows\System32\SETUPAPI.dll
C:\Windows\System32\COMDLG32.dll
C:\Windows\System32\cfgmgr32.dll
C:\Windows\System32\shcore.dll
C:\Windows\System32\bcrypt.dll
C:\Windows\System32\SHLWAPI.dll
C:\Program Files\OpenToonz\glut64.dll
C:\Program Files\OpenToonz\image.dll
C:\Program Files\OpenToonz\colorfx.dll
C:\Program Files\OpenToonz\tnztools.dll
C:\Program Files\OpenToonz\tfarm.dll
C:\Windows\System32\WS2_32.dll
C:\Program Files\OpenToonz\EDSDK.dll
C:\Program Files\OpenToonz\tnzstdfx.dll
C:\Program Files\OpenToonz\toonzqt.dll
C:\Program Files\OpenToonz\sound.dll
C:\Program Files\OpenToonz\Qt5Widgets.dll
C:\Program Files\OpenToonz\Qt5Core.dll
C:\Program Files\OpenToonz\Qt5Script.dll
C:\Program Files\OpenToonz\toonzlib.dll
C:\Program Files\OpenToonz\opencv_world451.dll
C:\Program Files\OpenToonz\tnzext.dll
C:\Program Files\OpenToonz\tnzbase.dll
C:\Program Files\OpenToonz\Qt5OpenGL.dll
C:\Program Files\OpenToonz\tnzcore.dll
C:\Windows\SYSTEM32\MPR.dll
C:\Windows\SYSTEM32\GLU32.dll
C:\Windows\SYSTEM32\dbghelp.dll
C:\Program Files\OpenToonz\Qt5Multimedia.dll
C:\Program Files\OpenToonz\MSVCP140.dll
C:\Windows\SYSTEM32\MFPlat.DLL
C:\Windows\SYSTEM32\MF.dll
C:\Program Files\OpenToonz\VCRUNTIME140.dll
C:\Program Files\OpenToonz\VCRUNTIME140_1.dll
C:\Program Files\OpenToonz\Qt5Network.dll
C:\Windows\System32\CRYPT32.dll
C:\Windows\SYSTEM32\WINSPOOL.DRV
C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_5.82.19041.1110_none_792d1c772443f647\COMCTL32.dll
C:\Windows\SYSTEM32\WINMM.dll
C:\Windows\SYSTEM32\MSVFW32.dll
C:\Windows\SYSTEM32\AVIFIL32.dll
C:\Program Files\OpenToonz\MSVCP140_1.dll
C:\Program Files\OpenToonz\libmypaint-1-4-0.dll
C:\Program Files\OpenToonz\Qt5Gui.dll
C:\Windows\SYSTEM32\VERSION.dll
C:\Windows\SYSTEM32\UxTheme.dll
C:\Windows\SYSTEM32\dwmapi.dll
C:\Windows\SYSTEM32\USERENV.dll
C:\Windows\SYSTEM32\NETAPI32.dll
C:\Program Files\OpenToonz\CONCRT140.dll
C:\Windows\SYSTEM32\DNSAPI.dll
C:\Windows\SYSTEM32\IPHLPAPI.DLL
C:\Program Files\OpenToonz\Qt5Svg.dll
C:\Windows\SYSTEM32\MSACM32.dll
C:\Program Files\OpenToonz\glew32.dll
C:\Windows\SYSTEM32\dxgi.dll
C:\Windows\SYSTEM32\d3d11.dll
C:\Windows\SYSTEM32\winmmbase.dll
C:\Windows\SYSTEM32\NETUTILS.DLL
C:\Windows\SYSTEM32\SRVCLI.DLL
C:\Program Files\OpenToonz\libintl-8.dll
C:\Windows\SYSTEM32\CRYPTBASE.DLL
C:\Program Files\OpenToonz\libjson-c-2.dll
C:\Program Files\OpenToonz\libiconv-2.dll
C:\Windows\SYSTEM32\dbgcore.DLL
C:\Windows\SYSTEM32\MFCORE.DLL
C:\Windows\SYSTEM32\powrprof.dll
C:\Windows\SYSTEM32\kernel.appcore.dll
C:\Windows\SYSTEM32\ksuser.dll
C:\Windows\System32\IMM32.DLL
C:\Windows\System32\NSI.dll
C:\Windows\System32\bcryptPrimitives.dll
C:\Windows\SYSTEM32\RTWorkQ.DLL
C:\Windows\SYSTEM32\UMPDC.dll
C:\Program Files\OpenToonz\platforms\qwindows.dll
C:\Windows\SYSTEM32\WTSAPI32.dll
C:\Windows\SYSTEM32\windows.storage.dll
C:\Windows\SYSTEM32\Wldp.dll
C:\Windows\SYSTEM32\profapi.dll
C:\Program Files\OpenToonz\styles\qwindowsvistastyle.dll
C:\Program Files\OpenToonz\iconengines\qsvgicon.dll
C:\Windows\system32\d3d9.dll
C:\Windows\SYSTEM32\msasn1.dll
C:\Windows\SYSTEM32\cryptnet.dll
C:\Windows\System32\WINTRUST.DLL
C:\Windows\System32\imagehlp.dll
C:\Windows\SYSTEM32\CRYPTSP.dll
C:\Windows\system32\rsaenh.dll
C:\Windows\System32\MSCTF.dll
C:\Windows\SYSTEM32\textinputframework.dll
C:\Windows\System32\CoreMessaging.dll
C:\Windows\System32\CoreUIComponents.dll
C:\Windows\SYSTEM32\ntmarta.dll
C:\Windows\SYSTEM32\wintypes.dll
C:\Windows\system32\Oleacc.dll
C:\Windows\system32\UIAutomationCore.dll
C:\Windows\SYSTEM32\PROPSYS.dll
C:\Windows\System32\clbcatq.dll
C:\Windows\SYSTEM32\sxs.dll
C:\Windows\System32\AppXDeploymentClient.dll
C:\Windows\System32\DriverStore\FileRepository\nv_dispig.inf_amd64_ee20464bb4ac57f4\nvoglv64.dll
C:\Windows\SYSTEM32\drvstore.dll
C:\Windows\SYSTEM32\devobj.dll
C:\Windows\SYSTEM32\dxcore.dll
C:\Windows\SYSTEM32\WINSTA.dll
C:\Windows\system32\dataexchange.dll
C:\Windows\system32\dcomp.dll
C:\Windows\system32\twinapi.appcore.dll
C:\Windows\system32\dwrite.dll
C:\Program Files\OpenToonz\imageformats\qgif.dll
C:\Program Files\OpenToonz\imageformats\qicns.dll
C:\Program Files\OpenToonz\imageformats\qico.dll
C:\Program Files\OpenToonz\imageformats\qjpeg.dll
C:\Program Files\OpenToonz\imageformats\qsvg.dll
C:\Program Files\OpenToonz\imageformats\qtga.dll
C:\Program Files\OpenToonz\imageformats\qtiff.dll
C:\Program Files\OpenToonz\imageformats\qwbmp.dll
C:\Program Files\OpenToonz\imageformats\qwebp.dll
C:\Program Files\OpenToonz\bearer\qgenericbearer.dll
C:\Windows\SYSTEM32\dhcpcsvc6.DLL
C:\Windows\SYSTEM32\dhcpcsvc.DLL
C:\Windows\System32\netprofm.dll
C:\Windows\System32\npmproxy.dll
C:\Program Files\OpenToonz\audio\qtaudio_windows.dll
C:\Windows\SYSTEM32\MMDevAPI.DLL
C:\Windows\SYSTEM32\wdmaud.drv
C:\Windows\SYSTEM32\AVRT.dll
C:\Windows\SYSTEM32\AUDIOSES.DLL
C:\Windows\SYSTEM32\msacm32.drv
C:\Windows\SYSTEM32\midimap.dll
C:\Windows\System32\mswsock.dll
C:\Windows\System32\rasadhlp.dll
C:\Windows\system32\napinsp.dll
C:\Windows\system32\pnrpnsp.dll
C:\Windows\system32\wshbth.dll
C:\Windows\system32\NLAapi.dll
C:\Windows\System32\winrnr.dll
C:\Windows\System32\fwpuclnt.dll
C:\Windows\SYSTEM32\LINKINFO.dll
C:\Windows\system32\wintab32.dll
C:\Windows\system32\winhttp.dll
C:\Windows\SYSTEM32\SspiCli.dll
C:\Windows\SYSTEM32\edputil.dll
C:\Windows\system32\Ninput.dll
C:\Windows\WinSxS\amd64_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.1110_none_60b5254171f9507e\comctl32.DLL
==== End ====

==== Backtrace ====
  1> RtlDeleteAce [0x7ffcdb368b4c] <ntdll.dll>
  2> RtlRaiseException [0x7ffcdb3412c6] <ntdll.dll>
  3> KiUserExceptionDispatcher [0x7ffcdb390f0e] <ntdll.dll>
  4> QByteArray::QByteArray [0x7ffc78d71ba6] <Qt5Core.dll>
  5> TFrameId::TFrameId [0x7ffc78762e92] <tnzcore.dll>
  6> TXshSimpleLevel::renumber [0x7ffc78c046f7] <toonzlib.dll>
  7> ToolUtils::TToolUndo::removeLevelAndFrameIfNeeded [0x7ffc8d6a2a91] <tnztools.dll>
  8> ToolUtils::TRasterUndo::undo [0x7ffc8d6a3be8] <tnztools.dll>
  9> TUndoManager::undo [0x7ffc7877dc63] <tnzcore.dll>
 10> TSmartPointerT<TXshSoundLevel>::~TSmartPointerT<TXshSoundLevel> [0x7ff7fb8da6c3] <OpenToonz.exe>
 11> QObject::qt_static_metacall [0x7ffc78f620d7] <Qt5Core.dll>
 12> QAction::activate [0x7ffc79345eff] <Qt5Widgets.dll>
 13> QAction::event [0x7ffc79346ebb] <Qt5Widgets.dll>
 14> QApplicationPrivate::notify_helper [0x7ffc793555c0] <Qt5Widgets.dll>
 15> QApplication::notify [0x7ffc79354643] <Qt5Widgets.dll>
 16> QCoreApplication::notifyInternal2 [0x7ffc78f4318a] <Qt5Core.dll>
 17> QShortcutMap::dispatchEvent [0x7ffc781296b2] <Qt5Gui.dll>
 18> QShortcutMap::tryShortcut [0x7ffc7812bb9a] <Qt5Gui.dll>
 19> QWindowSystemInterface::handleShortcutEvent [0x7ffc780edaef] <Qt5Gui.dll>
 20> QGuiApplicationPrivate::processKeyEvent [0x7ffc78103e7d] <Qt5Gui.dll>
 21> QWindowSystemInterface::sendWindowSystemEvents [0x7ffc780ef030] <Qt5Gui.dll>
 22> QEventDispatcherWin32::processEvents [0x7ffc78f8bcfa] <Qt5Core.dll>
 23> qt_plugin_query_metadata [0x7ffc8d5728d9] <qwindows.dll>
 24> QEventLoop::exec [0x7ffc78f3f64c] <Qt5Core.dll>
 25> QCoreApplication::exec [0x7ffc78f42164] <Qt5Core.dll>
 26> TSmartPointerT<TXshSoundLevel>::~TSmartPointerT<TXshSoundLevel> [0x7ff7fb8c5319] <OpenToonz.exe>
 27> TSmartPointerT<TXshSoundTextLevel>::~TSmartPointerT<TXshSoundTextLevel> [0x7ff7fbcf2367] <OpenToonz.exe>
 28> TSmartPointerT<TXshSoundTextLevel>::~TSmartPointerT<TXshSoundTextLevel> [0x7ff7fbcf1942] <OpenToonz.exe>
 29> BaseThreadInitThunk [0x7ffcd9e47614] <KERNEL32.DLL>
 30> RtlUserThreadStart [0x7ffcdb3426a1] <ntdll.dll>
==== End ====
![image](https://github.com/opentoonz/opentoonz/assets/97484115/c1e84d0e-d0d7-43af-9a1d-4652c04c51ab)

### OpenToonz Version

1.7

### OpenToonz Version Information

nightly build 6/8/2023

### Operating System

Windows

### GPU

NVIDIA

### Graphics Tablet

Huion

**Comment by JoeyEdmo** [comment]
@RodneyBaker @Bracket-H  thank you for the quick responses! 

I don't believe the levels initially had missing frames. So on the column 1-7 were filled. From what I remember it was using undo repetitively that caused the frames to go red and crash. Frame 7 also disappeared. I thought it was weird and included it in the screenshot. It's also interesting this happened both times it crashed, not sure how helpful it is though. 

drivers: I'll try tweaking/reinstalling drivers, but I don't believe I've change settings from defaults.

Workflow: I've just started toying around with opentoonz, so my workflow is far from complex. I was just doing frame by frame drawing in one column, and occasionally changing the animation frame timings (1s vs 2s vs 3s). This was all in the sandbox project with a new scene, settings unchanged from default. I stayed in the basic drawing room and just using the scene manager on bottom to create frames and whatnot. Wasn't uploading anything in or doing anything special, besides using alt-f and alt-9 to go full-screen occasionally. 

I couldn't tell you why there's 32 bit references in there, but the cpu is 64 bit. Also I may be blind but I don't see any 32 bit references, the backtrace seems to be using 48bit  (ex: 0x7ffc79345eff)  refs which I believe is expected. 

If I can isolate the bug in the future I'll make sure to update the issue. In the meantime I'll try saving more frequently and avoiding significant use of undo, or at least be more thoughtful when I use it.

Let me know if there's anything I missed, or any other info you'd like. Thanks again!
